module/os/linux/spl/spl-zone.c

```diff
@@ -281,13 +265,38 @@ zone_dataset_visible(const char *dataset, int *write)
 }
 EXPORT_SYMBOL(zone_dataset_visible);
 
+#if defined(CONFIG_USER_NS) && defined(HAVE_USER_NS_COMMON_INUM)
+static int
+user_ns_get(int fd, struct user_namespace **userns)
+{
+	struct file *nsfile;
+	struct ns_common *ns;
+	int error;
+
+	if ((nsfile = fget(fd)) == NULL)
+		return (EBADF);
+	ns = get_proc_ns(file_inode(nsfile));
+	if (ns->ops->type != CLONE_NEWUSER) {
```

Comment:
There needs to be an additional check that the file is safe to use.
(e.g. `zfs zone $(mktemp) pool/data` 
currently causes a kernel NULL pointer dereference).

Last time I checked the kernel has that kind of functionality but does not export it to modules.
Since you dropped pre 3.19 kernel support another way to check that the file is valid is to test for `NSFS_MAGIC`.
```
struct kstatfs st;
if (vfs_statfs(&file->f_path,&st) == 0 && st.f_type == NSFS_MAGIC)
```
But there are properly a lot more ways the achieve this.